beacon_node/http_api/src/attestation_performance.rs

```diff
@@ -0,0 +1,205 @@
+use beacon_chain::{BeaconChain, BeaconChainError, BeaconChainTypes};
+use eth2::lighthouse::{
+    AttestationPerformance, AttestationPerformanceQuery, AttestationPerformanceStatistics,
+};
+use state_processing::{
+    per_epoch_processing::altair::participation_cache::Error as ParticipationCacheError,
+    per_epoch_processing::EpochProcessingSummary, BlockReplayError, BlockReplayer,
+};
+use std::sync::Arc;
+use types::{BeaconState, BeaconStateError, EthSpec, Hash256, SignedBeaconBlock};
+use warp_utils::reject::{beacon_chain_error, custom_bad_request, custom_server_error};
+
+const MAX_REQUEST_RANGE_EPOCHS: usize = 100;
+const BLOCK_ROOT_CHUNK_SIZE: usize = 100;
+
+#[derive(Debug)]
+enum AttestationPerformanceError {
+    BlockReplay(BlockReplayError),
+    BeaconState(BeaconStateError),
+    ParticipationCache(ParticipationCacheError),
+    UnableToFindValidator(usize),
+}
+
+impl From<BlockReplayError> for AttestationPerformanceError {
+    fn from(e: BlockReplayError) -> Self {
+        Self::BlockReplay(e)
+    }
+}
+
+impl From<BeaconStateError> for AttestationPerformanceError {
+    fn from(e: BeaconStateError) -> Self {
+        Self::BeaconState(e)
+    }
+}
+
+impl From<ParticipationCacheError> for AttestationPerformanceError {
+    fn from(e: ParticipationCacheError) -> Self {
+        Self::ParticipationCache(e)
+    }
+}
+
+pub fn get_attestation_performance<T: BeaconChainTypes>(
+    target: String,
+    query: AttestationPerformanceQuery,
+    chain: Arc<BeaconChain<T>>,
+) -> Result<Vec<AttestationPerformance>, warp::Rejection> {
+    let spec = &chain.spec;
+    // We increment by 2 here so that when we build the state from the `prior_slot` it is
+    // still 1 epoch ahead of the first epoch we want to analyse.
+    // This ensures the `.is_previous_epoch_X` functions on `EpochProcessingSummary` return results
+    // for the correct epoch.
+    let start_epoch = query.start_epoch + 2;
+    let start_slot = start_epoch.start_slot(T::EthSpec::slots_per_epoch());
+    let prior_slot = start_slot - 1;
+
+    let end_epoch = query.end_epoch + 2;
+    let end_slot = end_epoch.end_slot(T::EthSpec::slots_per_epoch());
+
+    // Check query is valid
+    if start_epoch > end_epoch {
+        return Err(custom_bad_request(format!(
+            "invalid start and end epochs: {}, {}",
+            query.start_epoch, query.end_epoch
+        )));
+    }
```

Comment:
Sorry, me again :sweat_smile:

I noticed that we return a 500 internal server error if the `query.start_epoch` is `>= current_epoch - 1` due to the lookahead stuff. Would you mind adding a condition here like: `if query.end_epoch >= current_epoch - 1 { /* bad request */ }`. There are probably a few ways to write equivalent conditions but I think looking at the `end_epoch` makes sense. I think pretty much we need to just enforce that queries only include data up to the `current_epoch - 2`.

Reply:
Oh nice catch!
I think I've fixed it now. 
I also adjusted some of error messages